src/main/java/org/elasticsearch/plugins/PluginManager.java

```diff
@@ -19,674 +19,557 @@
-                //sometimes (e.g. github) the downloaded archive contains a top-level folder which needs to be removed
-                final boolean stripTopLevelDirectory;
-                if (topLevelFiles.length == 1 && Files.isDirectory(topLevelFiles[0])) {
-                    // valid names if the zip has only one top level directory
-                    switch (topLevelFiles[0].getFileName().toString()) {
-                        case  "_site/":
-                        case  "bin/":
-                        case  "config/":
-                        case  "_dict/":
-                          stripTopLevelDirectory = false;
-                          break;
-                        default:
-                          stripTopLevelDirectory = true;
+
+        @Override
+        public ExitStatus execute(Settings settings, Environment env) throws Exception {
+            final PluginDescriptor removable = pluginDescriptor(fqn).build();
+            validatePluginName(removable.name());
+            terminal.println("Uninstalling plugin [%s]...", removable.name());
+
+            // Lists all installed plugins with same name & version
+            PluginRepository local = installedPluginsRepository(env, settings);
+            Collection<PluginDescriptor> plugins = local.list(new PluginRepository.Filter() {
+                @Override
+                public boolean accept(PluginDescriptor plugin) {
+                    boolean match = StringUtils.equals(removable.name(), plugin.name());
+                    if (Strings.hasText(removable.version())) {
+                        return match && StringUtils.equals(removable.version(), plugin.version());
                     }
-                } else {
-                    stripTopLevelDirectory = false;
+                    return match;
                 }
-                Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
-                    @Override
-                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-                        Path target =  FileSystemUtils.append(extractLocation, file, stripTopLevelDirectory ? 1 : 0);
-                        Files.createDirectories(target);
-                        Files.copy(file, target, StandardCopyOption.REPLACE_EXISTING);
-                        return FileVisitResult.CONTINUE;
-                    }
+            });
 
-                });
-            }
-            log("Installed " + name + " into " + extractLocation.toAbsolutePath());
-        } catch (Exception e) {
-            log("failed to extract plugin [" + pluginFile + "]: " + ExceptionsHelper.detailedMessage(e));
-            return;
-        } finally {
-            try {
-                Files.delete(pluginFile);
-            } catch (Exception ex) {
-                log("Failed to delete plugin file" + pluginFile + " " + ex);
+            if ((plugins == null) || (plugins.isEmpty())) {
```

Comment:
Looking at the code, `plugins` should never be `null`. Not sure we need to test it.